tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Session.java

```diff
@@ -477,15 +544,49 @@ public Runner runner() {
    * @return if wantRunMetadata is true, serialized representation of the RunMetadata protocol
    *     buffer, false otherwise.
    */
-  private static native byte[] run(
-      long handle,
+  private static byte[] run(
+      TF_Session handle,
       byte[] runOptions,
-      long[] inputTensorHandles,
-      long[] inputOpHandles,
+      TF_Tensor[] inputTensorHandles,
+      TF_Operation[] inputOpHandles,
       int[] inputOpIndices,
-      long[] outputOpHandles,
+      TF_Operation[] outputOpHandles,
       int[] outputOpIndices,
-      long[] targetOpHandles,
+      TF_Operation[] targetOpHandles,
       boolean wantRunMetadata,
-      long[] outputTensorHandles);
+      TF_Tensor[] outputTensorHandles) {
+    requireHandle(handle);
+
+    int ninputs = inputTensorHandles.length;
+    int noutputs = outputTensorHandles.length;
+    int ntargets = targetOpHandles.length;
+
+    try (PointerScope scope = new PointerScope()) {
+      TF_Output inputs = new TF_Output(ninputs);
+      PointerPointer<TF_Tensor> inputValues = new PointerPointer<TF_Tensor>(ninputs);
+      TF_Output outputs = new TF_Output(noutputs);
+      PointerPointer<TF_Tensor> outputValues = new PointerPointer<TF_Tensor>(noutputs);
+      PointerPointer<TF_Operation> targets = new PointerPointer<TF_Operation>(ntargets);
+      TF_Buffer runMetadata = wantRunMetadata ? TF_Buffer.newBuffer() : null;
+
+      resolveHandles("input Tensors", inputTensorHandles, inputValues, ninputs);
+      resolveOutputs("input", inputOpHandles, inputOpIndices, inputs, ninputs);
+      resolveOutputs("output", outputOpHandles, outputOpIndices, outputs, noutputs);
+      resolveHandles("target Operations", targetOpHandles, targets, ntargets);
+
+      TF_Status status = TF_Status.newStatus();
+      TF_Buffer runOpts = TF_Buffer.newBufferFromString(runOptions);
+
+      TF_SessionRun(handle, runOpts, inputs, inputValues, ninputs,
+                    outputs, outputValues, noutputs, targets, ntargets,
+                    runMetadata, status);
+      status.throwExceptionIfNotOK();
+
+      for (int i = 0; i < noutputs; ++i) {
+        outputTensorHandles[i] = outputValues.get(TF_Tensor.class, i);
+      }
+
+      return runMetadata != null ? runMetadata.get() : null;
```

Comment:
Question: did you ever ran a benchmark showing that all these context switches between the JVM and the native code (i.e. at each JavaCPP generated method) ends up to be as performant as when only one call was made (`run` in this case)?

Reply:
Calls from the JVM to native code are very efficient, in the order of 30 ns. On the other hand, calls from native code to the JVM are very expensive, in the order of 300 ns, so it's almost certain that the new code here is going to faster. I'll run a simple benchmark and post the results here just to confirm, but if you're worried performance, we should think about writing a whole set of benchmarks to make sure there is never any regression in performance anywhere.

Reply:
I've started to write some of them as well (like [this one](https://github.com/tensorflow/java/blob/master/tensorflow-tools/src/test/java/org/tensorflow/tools/benchmark/NdArrayBenchmark.java)), I think with high-performance projects like TF we should have a set of benchmarks, not only it is useful to detect regression but also is a very good indicator of where we should spend time for optimization. 

If you can add some, that would be great, but we can merge without if you want to create these later.